File: backend/jobs/scheduler.py
"""
Configuração e gerenciamento do APScheduler
"""
import logging
import pytz
from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.cron import CronTrigger
from apscheduler.triggers.interval import IntervalTrigger
from apscheduler.executors.pool import ThreadPoolExecutor
from backend.jobs.sync_agendamentos import sync_agendamentos_job
from backend.jobs.remind_today import remind_today_job

logger = logging.getLogger(__name__)

# Timezone de São Paulo
SP_TZ = pytz.timezone('America/Sao_Paulo')

class JobScheduler:
    """Gerenciador de jobs automáticos"""
    
    def __init__(self):
        # Configurar executor com pool de threads
        executors = {
            'default': ThreadPoolExecutor(20),
        }
        
        # Configurar scheduler
        self.scheduler = BackgroundScheduler(
            executors=executors,
            timezone=SP_TZ,
            job_defaults={
                'coalesce': True,  # Combina execuções em atraso
                'max_instances': 1  # Apenas uma instância por job
            }
        )
        
        self._setup_jobs()
    
    def _setup_jobs(self):
        """Configura todos os jobs automáticos"""
        
        # Job 1: Sync agendamentos (a cada hora)
        self.scheduler.add_job(
            func=sync_agendamentos_job,
            trigger=IntervalTrigger(hours=1),
            id='sync_agendamentos',
            name='Sincronizar Agendamentos com Google Calendar',
            replace_existing=True,
            misfire_grace_time=300  # 5 minutos de tolerância
        )
        
        # Job 2: Lembrete diário (todo dia às 08:00)
        self.scheduler.add_job(
            func=remind_today_job,
            trigger=CronTrigger(hour=8, minute=0, timezone=SP_TZ),
            id='remind_today',
            name='Enviar Lembretes Diários',
            replace_existing=True,
            misfire_grace_time=1800  # 30 minutos de tolerância
        )
        
        logger.info("Jobs configurados com sucesso")
    
    def start(self):
        """Inicia o scheduler"""
        try:
            self.scheduler.start()
            logger.info("Scheduler iniciado com sucesso")
            self._log_scheduled_jobs()
        except Exception as e:
            logger.error(f"Erro ao iniciar scheduler: {e}")
    
    def shutdown(self):
        """Para o scheduler graciosamente"""
        try:
            self.scheduler.shutdown(wait=True)
            logger.info("Scheduler parado com sucesso")
        except Exception as e:
            logger.error(f"Erro ao parar scheduler: {e}")
    
    def _log_scheduled_jobs(self):
        """Log dos jobs agendados"""
        jobs = self.scheduler.get_jobs()
        logger.info(f"Jobs ativos: {len(jobs)}")
        
        for job in jobs:
            next_run = job.next_run_time
            if next_run:
                logger.info(f"  - {job.name}: próxima execução em {next_run.strftime('%d/%m/%Y %H:%M:%S %Z')}")
            else:
                logger.info(f"  - {job.name}: não agendado")
    
    def get_job_status(self) -> dict:
        """Retorna status de todos os jobs"""
        jobs = self.scheduler.get_jobs()
        
        status = {
            'scheduler_running': self.scheduler.running,
            'total_jobs': len(jobs),
            'jobs': []
        }
        
        for job in jobs:
            # APScheduler 4.x: next_run_time pode não estar disponível diretamente
            try:
                next_run = job.next_run_time
                next_run_str = next_run.isoformat() if next_run else None
            except AttributeError:
                # Fallback para versões mais recentes
                next_run_str = "Agendado"
            
            job_info = {
                'id': job.id,
                'name': job.name,
                'next_run': next_run_str,
                'trigger': str(job.trigger)
            }
            status['jobs'].append(job_info)
        
        return status
    
    def run_job_now(self, job_id: str) -> bool:
        """Executa um job imediatamente"""
        try:
            job = self.scheduler.get_job(job_id)
            if job:
                job.func()
                logger.info(f"Job {job_id} executado manualmente")
                return True
            else:
                logger.warning(f"Job {job_id} não encontrado")
                return False
        except Exception as e:
            logger.error(f"Erro ao executar job {job_id}: {e}")
            return False

# Instância global do scheduler
job_scheduler = JobScheduler()

File: server/services/scheduler/scheduler.ts
/**
 * Serviço de agendamento de jobs
 * Usa node-schedule para executar tarefas em horários específicos
 */
import schedule from 'node-schedule';
import { syncAppointmentsJob } from './jobs/sync-appointments.js';
import { dailyRemindersJob } from './jobs/daily-reminders.js';

const TIMEZONE = 'America/Sao_Paulo';

interface JobInfo {
  id: string;
  name: string;
  schedule: string;
  nextRun: Date | null;
  lastRun: Date | null;
  running: boolean;
}

class JobScheduler {
  private jobs: Map<string, schedule.Job> = new Map();
  private jobInfo: Map<string, { name: string; schedule: string; lastRun: Date | null }> = new Map();
  private started: boolean = false;

  /**
   * Inicia todos os jobs agendados
   */
  start(): void {
    if (this.started) {
      console.log('[SCHEDULER] Já iniciado');
      return;
    }

    console.log('[SCHEDULER] Iniciando scheduler de jobs...');

    // Job 1: Sincronização de agendamentos com Google Calendar (a cada hora, minuto 0)
    const syncJob = schedule.scheduleJob(
      { rule: '0 * * * *', tz: TIMEZONE },
      async () => {
        this.updateLastRun('sync_appointments');
        await syncAppointmentsJob();
      }
    );
    this.jobs.set('sync_appointments', syncJob);
    this.jobInfo.set('sync_appointments', {
      name: 'Sincronização Google Calendar',
      schedule: 'A cada hora (minuto 0)',
      lastRun: null,
    });

    // Job 2: Lembretes diários (às 08:00)
    const reminderJob = schedule.scheduleJob(
      { hour: 8, minute: 0, tz: TIMEZONE },
      async () => {
        this.updateLastRun('daily_reminders');
        await dailyRemindersJob();
      }
    );
    this.jobs.set('daily_reminders', reminderJob);
    this.jobInfo.set('daily_reminders', {
      name: 'Lembretes Diários',
      schedule: 'Todo dia às 08:00',
      lastRun: null,
    });

    this.started = true;
    console.log('[SCHEDULER] Jobs agendados:');
    this.logScheduledJobs();
  }

  /**
   * Encerra todos os jobs gracefully
   */
  async shutdown(): Promise<void> {
    console.log('[SCHEDULER] Encerrando scheduler...');
    await schedule.gracefulShutdown();
    this.jobs.clear();
    this.started = false;
    console.log('[SCHEDULER] Encerrado com sucesso');
  }

  /**
   * Retorna status de todos os jobs
   */
  getStatus(): { running: boolean; jobs: JobInfo[] } {
    const jobsInfo: JobInfo[] = [];

    for (const [id, job] of Array.from(this.jobs.entries())) {
      const info = this.jobInfo.get(id);
      jobsInfo.push({
        id,
        name: info?.name || id,
        schedule: info?.schedule || 'N/A',
        nextRun: job.nextInvocation(),
        lastRun: info?.lastRun || null,
        running: true,
      });
    }

    return {
      running: this.started,
      jobs: jobsInfo,
    };
  }

  /**
   * Executa um job manualmente
   */
  async runJobNow(jobId: string): Promise<boolean> {
    console.log(`[SCHEDULER] Executando job manualmente: ${jobId}`);

    switch (jobId) {
      case 'sync_appointments':
        this.updateLastRun(jobId);
        await syncAppointmentsJob();
        return true;

      case 'daily_reminders':
        this.updateLastRun(jobId);
        await dailyRemindersJob();
        return true;

      default:
        console.warn(`[SCHEDULER] Job desconhecido: ${jobId}`);
        return false;
    }
  }

  /**
   * Verifica se o scheduler está rodando
   */
  isRunning(): boolean {
    return this.started;
  }

  /**
   * Atualiza timestamp da última execução
   */
  private updateLastRun(jobId: string): void {
    const info = this.jobInfo.get(jobId);
    if (info) {
      info.lastRun = new Date();
    }
  }

  /**
   * Loga informações dos jobs agendados
   */
  private logScheduledJobs(): void {
    for (const [id, job] of Array.from(this.jobs.entries())) {
      const info = this.jobInfo.get(id);
      const next = job.nextInvocation();
      console.log(`  - ${info?.name || id}: próxima execução ${next?.toISOString() || 'não agendado'}`);
    }
  }
}

// Singleton
export const jobScheduler = new JobScheduler();

/**
 * Inicializa o scheduler (chamado no startup do servidor)
 */
export function initScheduler(): void {
  jobScheduler.start();
}

/**
 * Encerra o scheduler (chamado no shutdown do servidor)
 */
export async function shutdownScheduler(): Promise<void> {
  await jobScheduler.shutdown();
}
